consensus/babe/babe.go

```diff
@@ -40,3 +50,67 @@ func NewBabeSession(pubkey VrfPublicKey, privkey VrfPrivateKey, rt *runtime.Runt
 		rt:            rt,
 	}
 }
+
+// runs the slot lottery for a specific slot
+// returns true if validator is authorized to produce a block for that slot, false otherwise
+func (b *BabeSession) runLottery(slot uint64) (bool, error) {
+	if slot < b.epochData.StartSlot {
+		return false, errors.New("slot is not in this epoch")
+	}
+
+	output, err := b.vrfSign(slot)
+	if err != nil {
+		return false, err
+	}
+
+	output_int := new(big.Int).SetBytes(output)
+	threshold, err := calculateThreshold(b.config.C1, b.config.C2, b.authorityIndex, b.authorityWeights)
+	if err != nil {
+		return false, err
+	}
+
+	return output_int.Cmp(threshold) > 0, nil
+}
+
+func (b *BabeSession) vrfSign(slot uint64) ([]byte, error) {
+	// TOOD: return VRF output and proof
+	// sign b.epochData.Randomness and slot
+	out := make([]byte, 32)
+	_, err := rand.Read(out)
+	return out, err
+}
+
+// calculates the slot lottery threshold for the authority at authorityIndex.
+// equation: threshold = 2^128 * (1 - (1-c)^(w_k/sum(w_i)))
+// where k is the authority index, and sum(w_i) is the
+// sum of all the authority weights
+// see: https://github.com/paritytech/substrate/blob/master/core/consensus/babe/src/lib.rs#L1022
+func calculateThreshold(C1, C2, authorityIndex uint64, authorityWeights []uint64) (*big.Int, error) {
```

Comment:
Perhaps this should have `BabeSession` as receiver. What do you think?

Reply:
I'm not sure, I kinda like having it as a standalone function since it's just a math function. I also modeled it off the rust code if it makes a difference. what do you think?